### DISCUSSION TITLE
Beatmap freezes on gameplay start when running in single-threaded mode

### DISCUSSION THREAD

**Discussion by Axilotl17** [ppy/osu#17361] [Q&A] [answered]
Video explaining and showing bug:
https://youtu.be/tpljAB_gQDA

i think this pause happens on literally every map, you only notice this on maps where the first note is close to the start. my logic behind this is that the cursor goes dark when it pauses, and it goes dark at the start of every map.

[logs.zip](https://github.com/ppy/osu/files/8310771/logs.zip)

**Comment by frenzibyte** [comment]
Sounds like the game window is failing to take focus. Can you open the frame graph (Ctrl+F11) and double-check that? if it's not focused, it'll be capped to 60hz.

**Reply by Axilotl17** [reply]
nope stays at 960hz

https://user-images.githubusercontent.com/78441708/159175650-185d624a-21a3-41b9-ad94-4b512f845ad3.mp4

**Reply by frenzibyte** [reply]
I thought the issue was about the game actually going to pause menu at first, not that the beatmap itself freezes. You might need to reword the title to avoid confusion.

**Comment by bdach** [comment]
what linux distribution is this? are you using pipewire or similar unconventional audio setup?

**Reply by Axilotl17** [reply]
i believe i use pipewire, and im on Geruda Arch

**Reply by bdach** [reply]
are you able to try and check if the issue happens without pipewire?

**Reply by Axilotl17** [reply]
but this also happens on my IOS build testflight so im unsure if its my computer; i can check it on a macbook too

**Reply by bdach** [reply]
well i am personally not seeing similar issues on my machine. are you able to link some example maps that this happens on?

**Reply by Axilotl17** [reply]
https://osu.ppy.sh/beatmapsets/859608#osu/1799197
https://osu.ppy.sh/beatmapsets/1029668#osu/2153089
https://osu.ppy.sh/beatmapsets/68834#osu/202328
https://osu.ppy.sh/beatmapsets/807850#osu/1695382
https://osu.ppy.sh/beatmapsets/685622#osu/1451282

**Reply by Axilotl17** [reply]
> are you able to try and check if the issue happens without pipewire?

not really although i can try on my laptop

**Reply by Axilotl17** [reply]
<img width="631" alt="Screen Shot 2022-03-20 at 11 14 01 AM" src="https://user-images.githubusercontent.com/78441708/159176536-e205bd10-b49c-4cb5-85b0-678920c5a234.png">

well this might be its own issue xD

edit: nevermind i fixed this

**Reply by Axilotl17** [reply]
and it does not happen on my macos laptop! but i did check my IOS device with the testflight and it does happen there.

**Comment by peppy** [comment]
I think I've seen this happen on iOS once or twice. Potentially related to single thread runtime?

**Reply by frenzibyte** [reply]
It is indeed, there used to be a similar issue about it which has since been resolved (https://github.com/ppy/osu/issues/12541), in case that turns out to be relevant in some way.

**Comment by frenzibyte** [comment, marked as answer]
This boils down all the way to https://github.com/ppy/osu-framework/issues/4396. Firstly, only beatmaps with negative start times have this effect (e.g. https://osu.ppy.sh/beatmapsets/1218852#osu/2536716). https://github.com/ppy/osu-framework/issues/4396 causes `TrackBass` to appear as "running" during the first `DecoupleableInterpolatingFramedClock.ProcessFrame` call after the gameplay clock is seeked to a negative time, unlike in multi-threaded mode, where the track would be stopped.

Track being in a "running" state while in negative time causes it to behave incorrectly and use the source clock time instead of the decoupleable time, therefore causing the player to be in a freezing state until current time advances beyond the negative time zone.

Fixing https://github.com/ppy/osu-framework/issues/4396 will likely be similar to https://github.com/ppy/osu-framework/pull/4395. Will look into this.

**Reply by bdach** [reply]
i don't think that's a p0 issue but knock yourself out i guess

**Reply by frenzibyte** [reply]
Picked `priority:0` since it's causing beatmap intros to never play in single-threaded mode, but could be lowered down to `priority:1`, dunno.

**Reply by frenzibyte** [reply]
Should be resolved via ppy/osu-framework#5082.
